File: 3° semestre/Tecnicas-de-programacão-2/aula-24-08-30/Singleton/Singleton01.js
class Singleton{
    constructor(){
        if(Singleton.instance == null){
            Singleton.instance = this;
            this.value = 0;
        }
        return Singleton.instance;
    }

    increment(){
        this.value += 1;
        console.log(`Value: ${this.value}`);
    }
}

// cliente
const s1 = new Singleton(); // todas usam somente uma instancia, salvando processamento
const s2 = new Singleton();
const s3 = new Singleton();
const s4 = new Singleton();

console.log(s1 == s2); // true

s1.increment();
s2.increment();
s3.increment();
s4.increment();

File: 3° semestre/Tecnicas-de-programacão-2/aula-24-08-30/Singleton/Singleton01.py
class Singleton:
    instance = None

    def __new__(cls): # cls é palavra reservada
        if cls.instance is None:
            cls.instance = super().__new__(cls) # __init__ e __new__ são construtores
            cls.instance.value = 0
        return cls.instance
    
    def increment(self):
        self.value += 1
        print(f'value: {self.value}')

# cliente
s1 = Singleton()
s2 = Singleton()
s3 = Singleton()
s4 = Singleton()

print (s1 == s2) # true

s1.increment()
s2.increment()
s3.increment()
s4.increment()